shadow/src/Nixfmt/Pretty.hs

```diff
@@ -0,0 +1,362 @@
+    pretty (IDSelector i)              = pretty i
+    pretty (InterpolSelector interpol) = pretty interpol
+    pretty (StringSelector s)          = pretty s
+
+instance Pretty Selector where
+    pretty (Selector dot sel Nothing)
+        = pretty dot <> pretty sel
+
+    pretty (Selector dot sel (Just (kw, def)))
+        = pretty dot <> pretty sel
+          <> hardspace <> pretty kw <> hardspace <> pretty def
+
+instance Pretty Binder where
+    pretty (Inherit inherit Nothing ids semicolon)
+        = group (pretty inherit <> softline
+                 <> nest 2 (sepBy softline ids)) <> pretty semicolon
+
+    pretty (Inherit inherit source ids semicolon)
+        = group (pretty inherit <> hardspace
+                 <> pretty source <> line
+                 <> nest 2 (sepBy softline ids)) <> pretty semicolon
+
+    pretty (Assignment selectors assign
+            (Term (String s@(Ann [[TextPart t]] Nothing []))) semicolon)
+        | validURI t = group (hcat selectors <> hardspace
+            <> pretty assign <> line
+            <> nest 2 (pretty t)) <> pretty semicolon
+        | otherwise = group (hcat selectors <> hardspace
+            <> pretty assign <> line
+            <> nest 2 (pretty s)) <> pretty semicolon
+
+    pretty (Assignment selectors assign expr semicolon)
+        = group (hcat selectors <> hardspace
+            <> pretty assign <> softline
+            <> pretty expr) <> pretty semicolon
+
+    pretty (BinderTrivia trivia) = pretty trivia
+
+instance Pretty ListPart where
+    pretty (ListItem term)     = pretty term
+    pretty (ListTrivia trivia) = pretty trivia
+
+-- | Pretty print a term without wrapping it in a group.
+prettyTerm :: Term -> Doc
+prettyTerm (List (Ann paropen Nothing []) [] parclose)
+    = pretty paropen <> pretty parclose
+
+prettyTerm (List (Ann paropen Nothing []) [item] parclose)
+    = pretty paropen <> pretty item <> pretty parclose
+
+prettyTerm (List (Ann paropen trailing leading) items parclose)
+    = pretty paropen <> pretty trailing <> line
+        <> nest 2 (sepBy line items') <> line
+        <> pretty parclose
+    where items' = case leading of
+                        [] -> items
+                        _  -> ListTrivia leading : items
+
+prettyTerm (Set Nothing (Ann paropen Nothing []) [] parclose)
+    = pretty paropen <> hardspace <> pretty parclose
+
+prettyTerm (Set Nothing (Ann paropen Nothing []) [item] parclose)
+    = pretty paropen <> line <> nest 2 (pretty item) <> line <> pretty parclose
+
+prettyTerm (Set krec (Ann paropen trailing leading) binders parclose)
+    = pretty (fmap ((<>hardspace) . pretty) krec)
+        <> pretty paropen <> pretty trailing <> line
+        <> nest 2 (sepBy hardline binders') <> line
+        <> pretty parclose
+    where binders' = case leading of
+                          [] -> binders
+                          _  -> BinderTrivia leading : binders
+
+prettyTerm (String x) = pretty x
+
+prettyTerm _ = undefined
```

Comment:
It's better to use `error` and write at least something that can give a hint where a bug happened if it even happens. Also consider adding `HasCallStack` constraint to this function.